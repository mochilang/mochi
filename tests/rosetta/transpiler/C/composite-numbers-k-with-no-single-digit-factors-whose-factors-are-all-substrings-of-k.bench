{
  "duration_us": 657,
  "memory_bytes": 1732608,
  "name": "main"
}