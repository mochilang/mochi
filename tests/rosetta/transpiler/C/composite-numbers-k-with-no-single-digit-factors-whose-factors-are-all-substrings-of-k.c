// Generated by Mochi 0.10.32 on 2025-08-05 01:12 +0700
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <malloc.h>

size_t primeFactors_len;

size_t _slice_int_len;
static char* str_concat(const char *a, const char *b) {
    size_t len1 = strlen(a);
    size_t len2 = strlen(b);
    char *res = malloc(len1 + len2 + 1);
    memcpy(res, a, len1);
    memcpy(res + len1, b, len2);
    res[len1 + len2] = 0;
    return res;
}

static char* str_int(long long v) {
    char buf[32];
    snprintf(buf, sizeof(buf), "%lld", v);
    return strdup(buf);
}

#include <time.h>
#include <stdlib.h>
static int seeded_now = 0;
static long long now_seed = 0;
static long long _now(void) {
    if (!seeded_now) {
        const char *s = getenv("MOCHI_NOW_SEED");
        if (s && *s) {
            now_seed = atoll(s);
            seeded_now = 1;
        }
    }
    if (seeded_now) {
        now_seed = (now_seed * 1664525 + 1013904223) % 2147483647;
        return now_seed;
    }
    struct timespec ts;
    clock_gettime(CLOCK_REALTIME, &ts);
    return (long long)(ts.tv_sec * 1000000000LL + ts.tv_nsec);
}

static long long _mem(void) {
    long long size = 0, rss = 0;
    FILE *f = fopen("/proc/self/statm", "r");
    if (f) {
        if (fscanf(f, "%lld %lld", &size, &rss) != 2) rss = 0;
        fclose(f);
    }
    return rss * (long long)sysconf(_SC_PAGESIZE);
}

static const char* _substring(const char *s, int start, int end) {
    int len = (int)strlen(s);
    if (start < 0) start = 0;
    if (end > len) end = len;
    if (start > end) start = end;
    char *res = malloc(end - start + 1);
    memcpy(res, s + start, end - start);
    res[end - start] = 0;
    return res;
}

static long long* _slice_int(const long long *arr, size_t len, int start, int end, size_t *out_len) {
    if (start < 0) start = 0;
    if ((size_t)end > len) end = len;
    if (start > end) start = end;
    size_t n = end - start;
    long long *res = NULL;
    if (n) {
        res = malloc(n * sizeof(long long));
        memcpy(res, arr + start, n * sizeof(long long));
    }
    *out_len = n;
    _slice_int_len = n;
    return res;
}

static long long* list_append_long_long(long long *arr, size_t *len, long long val) {
    arr = realloc(arr, (*len + 1) * sizeof(long long));
    arr[*len] = val;
    (*len)++;
    return arr;
}

long long * primeFactors(long long n);
const char* commatize(long long n);
long long indexOf(const char* s, const char* sub);
const char* pad10(const char* s);
const char* trimRightStr(const char* s);
void user_main();
int main(void);

long long * primeFactors(long long n) {
    long long *factors = NULL;
    size_t factors_len = 0;
    long long x = n;
    while ((x % 2LL) == 0LL) {
        factors = list_append_long_long(factors, &factors_len, 2LL);
        x = x / 2LL;
    }
    long long p = 3LL;
    while ((p * p) <= x) {
        while ((x % p) == 0LL) {
            factors = list_append_long_long(factors, &factors_len, p);
            x = x / p;
        }
        p = p + 2LL;
    }
    if (x > 1LL) {
        factors = list_append_long_long(factors, &factors_len, x);
    }
    return primeFactors_len = factors_len, factors;
}

const char* commatize(long long n) {
    const char* s = str_int(n);
    const char* out = "";
    long long i = strlen(s) - 1LL;
    long long c = 0LL;
    while (i >= 0LL) {
        out = str_concat(_substring(s, i, i + 1LL), out);
        c = c + 1LL;
        if (((c % 3LL) == 0LL) && (i > 0LL)) {
            out = str_concat(",", out);
        }
        i = i - 1LL;
    }
    return out;
}

long long indexOf(const char* s, const char* sub) {
    long long i = 0LL;
    while ((i + strlen(sub)) <= strlen(s)) {
        if (strcmp(_substring(s, i, i + strlen(sub)), sub) == 0) {
            return i;
        }
        i = i + 1LL;
    }
    return -1LL;
}

const char* pad10(const char* s) {
    const char* str = s;
    while (strlen(str) < 10LL) {
        str = str_concat(" ", str);
    }
    return str;
}

const char* trimRightStr(const char* s) {
    long long end = strlen(s);
    while ((end > 0LL) && (strcmp(_substring(s, end - 1LL, end), " ") == 0)) {
        end = end - 1LL;
    }
    return _substring(s, 0LL, end);
}

void user_main() {
    long long *res = NULL;
    size_t res_len = 0;
    long long count = 0LL;
    long long k = 121LL;
    while (count < 20LL) {
        if ((((k % 3LL) == 0LL) || ((k % 5LL) == 0LL)) || ((k % 7LL) == 0LL)) {
            k = k + 2LL;
            continue;
        }
        long long *factors = primeFactors(k);
        size_t factors_len = primeFactors_len;
        if (1LL) {
            const char* s = str_int(k);
            long long includesAll = 1LL;
            long long prev = -1LL;
            {
                long long f_arr[] = {11LL, 11LL};
                size_t f_len = sizeof(f_arr) / sizeof(f_arr[0]);
                for (size_t __i = 0; __i < f_len; __i++) {
                    long long f = f_arr[__i];
                    if (f == prev) {
                        continue;
                    }
                    const char* fs = str_int(f);
                    if (indexOf(s, fs) == -1LL) {
                        includesAll = 0LL;
                        break;
                    }
                    prev = f;
                }
            }
            if (includesAll) {
                res = list_append_long_long(res, &res_len, k);
                count = count + 1LL;
            }
        }
        k = k + 2LL;
    }
    const char* line = "";
    {
        long long* __tmp0 = _slice_int(res, res_len, 0LL, 10LL, &(_slice_int_len));
        size_t __tmp0_len = _slice_int_len;
        for (size_t __i = 0; __i < __tmp0_len; __i++) {
            long long e = __tmp0[__i];
            line = str_concat(str_concat(line, pad10(commatize(e))), " ");
        }
    }
    puts(trimRightStr(line));
    line = "";
    {
        long long* __tmp1 = _slice_int(res, res_len, 10LL, 20LL, &(_slice_int_len));
        size_t __tmp1_len = _slice_int_len;
        for (size_t __i = 0; __i < __tmp1_len; __i++) {
            long long e = __tmp1[__i];
            line = str_concat(str_concat(line, pad10(commatize(e))), " ");
        }
    }
    puts(trimRightStr(line));
}

int main(void) {
    {
        long long __start = _now();
        user_main();
        long long __end = _now();
        long long __dur_us = (__end - __start) / 1000;
        long long __mem_bytes = _mem();
        printf("{\n  \"duration_us\": %-lld,\n  \"memory_bytes\": %-lld,\n  \"name\": \"main\"\n}\n", __dur_us, __mem_bytes);
    }
    return 0;
}
