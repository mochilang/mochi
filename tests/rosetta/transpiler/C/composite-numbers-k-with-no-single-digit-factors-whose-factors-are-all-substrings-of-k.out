211        311        611        811        911      1,103      1,109      1,111      1,117      1,121
     1,123      1,129      1,133      1,139      1,147      1,151      1,153      1,157      1,159      1,163